### PR TITLE
Upgrade OpenShift with verification steps

### PR DESCRIPTION
**Pull request by ebattat**
## Type of change
Note: Fill **x** in []
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] dependencies

## Description

Upgrade OpenShift Container Platform (OCP) with verification steps:
There are 2 options: run_bare_metal_ocp_upgrade/ verify_bare_metal_upgrade_complete
1. run_bare_metal_ocp_upgrade:
    - Upgrade OCP using patch for OCP 4.16, need to run upgrade steps twice (sometimes the first run didnt work)
    - Adding waiting mechanism till upgrade start/end
    - Stop/Resume healthcheck before and after upgrade
    - Upgrade ODF: need to run Subscription with the upgrade channel version after upgrading
    - Adding sleep between upgrade command solved upgrade stuck issue:   
   
           ```
           oc adm upgrade
           sleep 10
           oc adm upgrade channel stable-4.15
           sleep 10
           oc adm upgrade --to=4.15.22
           ```

2. verify_bare_metal_upgrade_complete
    - Verify that OCP upgrade to target version
    - Verify that LSO operator upgrade to target version
    - Verify that ODF operator upgrade to target version
    - Verify that CNV operator upgrade to target version

### CODE REVIEW

**Comment by ebattat** [comment]
@RobertKrawitz, do you have any comment ?

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py [hunk 1]
```diff
@@ -92,6 +93,24 @@ def _wait_for_install_complete(self, sleep_time: int = SHORT_TIMEOUT):
             current_wait_time += sleep_time
         raise OCPInstallationFailed(install_log)
 
+    def _wait_for_upgrade_complete(self, oc: OC, sleep_time: int = SHORT_TIMEOUT):
```
Comment (RobertKrawitz) [inline benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:96]: How long is the short timeout?

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py [hunk 2]
```diff
@@ -215,18 +230,64 @@ def run_ocp_installer(self):
...
+        """
+        # check cluster version upgrade
+        if oc.get_upgrade_version() == self._upgrade_ocp_version:
+            lso_upgraded = oc.wait_for_operator_installation(operator='lso', version=lso_version, namespace='openshift-local-storage')
+            odf_upgraded = oc.wait_for_operator_installation(operator='odf', version=odf_version, namespace='openshift-storage')
+            cnv_upgraded = oc.wait_for_operator_installation(operator='cnv', version=cnv_version, namespace='openshift-cnv')
+            # check operators versions upgrade
+            if cnv_upgraded and lso_upgraded and odf_upgraded:
+                return True
+        return False
```
Comment (RobertKrawitz) [inline benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:278]: Not having all of the operators upgraded (yet) is a different condition than the desired upgrade version being wrong.  The former looks like a condition that may eventually resolve itself.  The latter is not going to be resolved and should probably raise an exception of some kind.

File: benchmark_runner/common/oc/oc.py [hunk 3]
```diff
@@ -48,6 +49,104 @@ def get_ocp_server_version(self):
...
+        @return:
+        """
+        ocp_channel = '.'.join(upgrade_ocp_version.split('.')[:2])
+        upgrade_command = f"{self.__cli} adm upgrade ; {self.__cli} adm upgrade channel stable-{ocp_channel} ; {self.__cli} adm upgrade --to={upgrade_ocp_version} --allow-explicit-upgrade;"
+
+        if ocp_channel == "4.16":
+            patch_command = f"{self.__cli} -n openshift-config patch cm admin-acks --patch '{{\"data\":{{\"ack-4.15-kube-1.29-api-removals-in-4.16\":\"true\"}}}}' --type=merge"
+            self.run(patch_command)
+
+        # sometimes upgrade is not working on the first run
```
Comment (RobertKrawitz) [inline benchmark_runner/common/oc/oc.py:66]: What goes wrong the first time?

File: benchmark_runner/common/oc/oc.py [hunk 4]
```diff
@@ -48,6 +49,104 @@ def get_ocp_server_version(self):
...
+        """
+        This method upgrades OCP version with conditional handling for specific versions.
+
+        @param upgrade_ocp_version: Version to upgrade to
+        @return:
+        """
+        ocp_channel = '.'.join(upgrade_ocp_version.split('.')[:2])
+        upgrade_command = f"{self.__cli} adm upgrade ; {self.__cli} adm upgrade channel stable-{ocp_channel} ; {self.__cli} adm upgrade --to={upgrade_ocp_version} --allow-explicit-upgrade;"
+
+        if ocp_channel == "4.16":
```
Comment (RobertKrawitz) [inline benchmark_runner/common/oc/oc.py:62]: Is this really a 4.16 flag day, or are you likely to have things like this happen with other releases?

**Review by RobertKrawitz — CHANGES_REQUESTED** [review]
Sorry, I missed the original request for review.  Just a few things.

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py (hunk 1 above)
Comment (ebattat) [reply benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:96]: [600 sec](https://github.com/redhat-performance/benchmark-runner/blob/a1fef4f7b6779e033c464f36933389cd5ef84c20/benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py#L23)

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py (hunk 2 above)
Comment (ebattat) [reply benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:278]: Right, add RuntimeError error:
```
        if bare_metal_operations.is_cluster_upgraded(oc, cnv_version=cnv_version, odf_version=odf_version, lso_version=lso_version):
            bare_metal_operations.verify_cluster_is_up(oc)
        else:
            error_message = f'OCP {upgrade_ocp_version} upgrade failed'
            logger.error(error_message)
            raise RuntimeError(error_message)
```

File: benchmark_runner/common/oc/oc.py (hunk 3 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py:66]: Seems that the upgrade did not start

File: benchmark_runner/common/oc/oc.py (hunk 4 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py:62]: Yes, it special instruction for ocp 4.16.
add comment https://access.redhat.com/articles/7031404

File: benchmark_runner/common/oc/oc.py (hunk 3 above)
Comment (ebattat) [reply benchmark_runner/common/oc/oc.py]: Change the comment to:
# The upgrade is triggered only when it is run twice

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py (hunk 1 above)
Comment (RobertKrawitz) [reply benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:96]: Think that will be long enough?

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py (hunk 2 above)
Comment (RobertKrawitz) [reply benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:278]: OK

File: benchmark_runner/common/oc/oc.py (hunk 4 above)
Comment (RobertKrawitz) [reply benchmark_runner/common/oc/oc.py:63]: OK

**Review by RobertKrawitz — APPROVED** [review]

File: benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py (hunk 1 above)
Comment (ebattat) [reply benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py:96]: Yes, its just the sleep wait in the loop.
The total timeout is [10800](https://github.com/redhat-performance/benchmark-runner/blob/4d50568db561651aa9836b6f3816e81b5da826d6/benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py#L84)

**Pull request merged**